src/circuit/gadgets/constraints.rs

```diff
@@ -698,6 +799,172 @@ mod tests {
 
     proptest! {
 
+        #[test]
+        fn test_enforce_equal((a, b) in any::<(FWrap<Fr>, FWrap<Fr>)>()) {
+            prop_assume!(a != b);
+
+            let test_a_b = |a, b| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let a_num = AllocatedNum::alloc(cs.namespace(|| "a_num"), || Ok(a)).unwrap();
+                let b_num = AllocatedNum::alloc(cs.namespace(|| "b_num"), || Ok(b)).unwrap();
+                let _ = enforce_equal(&mut cs, || "enforce equal", &a_num, &b_num);
+                assert_eq!(cs.is_satisfied(), a==b);
+            };
+
+            // positive
+            test_a_b(a.0, a.0);
+
+            // negative
+            test_a_b(a.0, b.0);
```

Comment:
Why are we performing the same check twice and labeling it as both positive and negative. I think we're actually only ever checking the negative case. As I mentioned in comments to a previous PR prototyping proptest in circuits, this doesn't seem to capture the goal well.

This is perfectly well tested with just two cases — one where the numbers are equal and one where they are not. Yet by using proptest, we've managed to accidentally write a test where one of the critical cases is never exercised.

I think as a group (those who might write these proptests) we need to stop and *think* about the implications of selecting random field elements. If we really want to do this, we need to stop and write strategies that will allow creating distributions of multiple field elements that yield the properties we intend to test in appropriate frequencies.

This is really important. Property tests *could* lead to a much more reliable system and better test coverage, but we *must* make sure that we're also clearly, obviously, and explicitly covering the cases that actually need to be tested (and with minimal overhead).

My concern right now is that we confuse ourselves into more complicated, more expensive test coverage that's actually less exhaustive. If we can't resolve it in a straightforward way — I'd suggest double coverage. Let's just write simple tests that definitely cover, then prop tests that spam data if we think that will help.

I might also be missing the point; and I don't yet have a developed theory about the right strategies that would be reusable in a way that reduces the amount of deep thought required for each test. My great fear is that if we can't arrive at such a strategy, then the need to carefully audit and reason about the tests *themselves* introduces a new vector for error.

Reply:
Positive test verifies that if the first input is equal to the second, then the constraints are satisfied. Then I passed the same value, `a.0` and `a.0` again. 

Negative test verifies that if the first input is **not** equal to the second, then the constraints are **not** satisfied. Then I passed different values `a.0`and `b.0`.

Reply:
Yep, as above and noted in my update — I simply read this wrong. This is my fault but also points to an ergonomic issue with tests like this. I think maybe longer and more obviously different *names* would help. This is not an indictment of the original test, just a further observation that the soup of characters makes the intent harder to observe than it should be.

For clarity, this is fine as written. I'm musing about how to improve readability.